src/qibolab/_core/instruments/qcs.py

```diff
@@ -0,0 +1,343 @@
+                        program.add_acquisition(duration, virtual_channel)
+
+                    elif pulse.kind == "readout":
+                        sweep_param_map = sweeper_pulse_map.get(pulse.probe.id, {})
+                        qcs_pulse = generate_qcs_rfwaveform(
+                            duration=sweep_param_map.get(
+                                "duration", pulse.probe.duration * NANOSECONDS
+                            ),
+                            envelope=pulse.probe.envelope,
+                            amplitude=sweep_param_map.get(
+                                "amplitude", pulse.probe.amplitude
+                            ),
+                            frequency=frequency,
+                            phase=sweep_param_map.get(
+                                "relative_phase", pulse.probe.relative_phase
+                            ),
+                        )
+                        integration_filter = qcs.IntegrationFilter(qcs_pulse)
+                        program.add_waveform(qcs_pulse, probe_virtual_channel)
+                        program.add_acquisition(integration_filter, virtual_channel)
+
+            elif isinstance(channel, IqChannel):
+
+                pulses = []
+                if channel_id in sweeper_channel_map:
+                    frequency = sweeper_channel_map[channel_id]
+                else:
+                    frequency = configs[channel_id].frequency
+
+                for pulse in sequence.channel(channel_id):
+                    sweep_param_map = sweeper_pulse_map.get(pulse.id, {})
+
+                    if pulse.kind == "delay":
+                        qcs_pulse = qcs.Delay(
+                            sweep_param_map.get(
+                                "duration", pulse.duration * NANOSECONDS
+                            )
+                        )
+                    elif pulse.kind == "virtualz":
+                        qcs_pulse = qcs.PhaseIncrement(
+                            phase=sweep_param_map.get("relative_phase", pulse.phase)
+                        )
+                    elif pulse.kind == "pulse":
+                        qcs_pulse = generate_qcs_rfwaveform(
+                            duration=sweep_param_map.get(
+                                "duration", pulse.duration * NANOSECONDS
+                            ),
+                            envelope=pulse.envelope,
+                            amplitude=sweep_param_map.get("amplitude", pulse.amplitude),
+                            frequency=frequency,
+                            phase=sweep_param_map.get(
+                                "relative_phase", pulse.relative_phase
+                            ),
+                        )
+                        if pulse.envelope.kind == "drag":
+                            qcs_pulse = qcs_pulse.drag(coeff=pulse.envelope.beta)
+                    else:
+                        raise ValueError("Unrecognized pulse type", pulse.kind)
```

Comment:
Just to sigh together: Python now has a `match` statement that would perfectly fits in here
https://docs.python.org/3/reference/compound_stmts.html#the-match-statement

But it has been introduced in py3.10. And we're still supporting py3.9 for one more year...